### ISSUE TITLE
Multiple scopes on server for webapi sync

### ISSUE THREAD

**Issue by mdonogma** [Mimetis/Dotmim.Sync#227]
The example in your docs explain creating scopes client side.

However with webapi sync we need scopes defined server side . That way filters, scopes etc are controlled by server

Is this possible ?

**Comment by Mimetis** [comment]
When you're using web api, the scope is defined on the server side, when configuring services:


``` cs
public void ConfigureServices(IServiceCollection services)
{
    services.AddMvc().SetCompatibilityVersion(CompatibilityVersion.Version_2_2);

    // [Required]: Handling multiple sessions
    services.AddMemoryCache();

    // [Required]: Get a connection string to your server data source
    var connectionString = Configuration.GetSection("ConnectionStrings")["DefaultConnection"];

    // [Optional]: Web server Options. Batching directory and Snapshots directory
    var options = new WebServerOptions()
    {
        BatchDirectory = Path.Combine(SyncOptions.GetDefaultUserBatchDiretory(), "server"),
        SnapshotsDirectory = Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.UserProfile), "Snapshots")
    };

    // [Required]: Tables list involved in the sync process
    var tables = new string[] {"ProductCategory",
                     "ProductDescription", "ProductModel",
                     "Product", "ProductModelProductDescription",
                     "Address", "Customer", "CustomerAddress",
                     "SalesOrderHeader", "SalesOrderDetail" };

    // [Optional]: database setup, for objects naming conventions
    var setup = new SyncSetup(tables)
    {
        // optional :
        StoredProceduresPrefix = "s",
        StoredProceduresSuffix = "",
        TrackingTablesPrefix = "s",
        TrackingTablesSuffix = ""
    };

    // [Required]: Add a SqlSyncProvider acting as the server hub.
    // If you don't specify a SyncSetup object, just add the tables array.
    services.AddSyncServer<SqlSyncProvider>(connectionString, setup, options);
}
```
 The client side saves this scope as a *cache* system so he does not have to get it back from server, afterwards.

Make sense ?

**Comment by mdonogma** [comment]
Your example has only 1 scope. 

How do you create 2 or more scopes server side then specify on client which scope it is synchronising with?

Example: 3 scopes on server.

On login sync with scope 1
On opening UI abc - sync scope 2
On opening UI ggg - sync scope 3

**Comment by Mimetis** [comment]
ok got it, let me think about it

**Comment by lordofffarm** [comment]
@Mimetis In my app I added two classes called Database and SyncServer.
```
    public class Database
    {
        public string Id { get; set; }
        public string Type { get; set; } // mysql
        public string ConnectionString { get; set; }
        public SyncConfiguration SyncConfiguration { get; set; }
    }

    public class SyncServer
    {
        public string SyncServerName { get; set; }
        public string DatabaseId { get; set; }
        public Database Database { get; set; }
        public IList<string> TablesToSync { get; set; }
    }
```

Database holds the SyncConfiguration which has all the tables that CAN be synced. SyncServer can have some or all of the tables from the SyncConfiguration.
When clients make requests to sync they send the SyncServerName string to the server.
When synchronizing, I intercept the event with DatabaseProvisioningArgs and remove the unnecessary tables.
An important thing is that Provisioning for the first time has to use the SyncConfiguration which has all the tables, otherwise if two SyncServers have the same tables, and the db has not yet been provisioned, the framework will try to provision those same tables and fail saying triggers/SPs already exist. That is for v0.3. Haven't tried the new version yet.

This approach also enabled me to use multiple databases in the same app.

**Comment by mdonogma** [comment]
Is it possible to build this into 4.x . 

Prefer something the framework supports in order to avoid future versions of dot in sync breaking it.

Example: client sends the list of tables and optional dB name when making sync requests and server side code knows how to deal with it. 

The sync options in the docs indicate sync options allow client or server to deduce what tables to sync but that is not really the case when using webapi. So docs probably need to be clear about that going forward. That way it saves developers finding out the hard way syncoptions don’t work in webapi scenario.

**Comment by mdonogma** [comment]
Can you share a copy of your DatabaseProvisioningArgs

**Comment by lordofffarm** [comment]
@mdonogma here is a sample
So when synchronizing I use SchemaArgs and DatabaseProvisioningArgs when provisioning.
This example is for v0.3. It might be different for the latest version.

```
    [Route("api/[controller]")]
    public class SyncController : ControllerBase
    {
        private readonly WebProxyServerProvider _webProxyServerProvider;

        public SyncController(WebProxyServerProvider webProxyServerProvider)
        {
            _webProxyServerProvider = webProxyServerProvider;
        }

        [HttpPost]
        public async Task Do()
        {
            var interceptor = new Interceptor<SchemaArgs>(args =>
            {
                IList<string> neededTables = Tables.GetNeededTables();
                var tablesToRemove = new List<DmTable>();
                // remove unnecessary tables
                foreach (var dmTable in schema.Tables)
                {
                    var any = neededTables.Any(table1 => table1 == dmTable.TableName);
                    if (any) continue;

                    foreach (var relation in dmTable.ParentRelations)
                    {
                        schema.Relations.Remove(relation);
                    }

                    foreach (var relation in dmTable.ChildRelations)
                    {
                        schema.Relations.Remove(relation);
                    }

                    tablesToRemove.Add(dmTable);
                }

                foreach (var dmTable in tablesToRemove)
                {
                    schema.Tables.Remove(dmTable);
                }
            });
            _webProxyServerProvider.GetLocalProvider(HttpContext).On(interceptor);

            await _webProxyServerProvider.HandleRequestAsync(HttpContext);
        }
    }
```

**Comment by mdonogma** [comment]
Any thought on supporting multiple scopes for webapi sync in 0.5 version.

Sent from my iPhone

On 27 Feb 2020, at 15:48, lordofffarm <notifications@github.com> wrote:

﻿

@mdonogma<https://github.com/mdonogma> here is a sample
So when synchronizing I use SchemaArgs and DatabaseProvisioningArgs when provisioning.
This example is for v0.3. It might be different for the latest version.

    [Route("api/[controller]")]
    public class SyncController : ControllerBase
    {
        private readonly WebProxyServerProvider _webProxyServerProvider;

        public SyncController(WebProxyServerProvider webProxyServerProvider)
        {
            _webProxyServerProvider = webProxyServerProvider;
        }

        [HttpPost]
        public async Task Do()
        {
            var interceptor = new Interceptor<SchemaArgs>(args =>
            {
                IList<string> neededTables = Tables.GetNeededTables();
                var tablesToRemove = new List<DmTable>();
                // remove unnecessary tables
                foreach (var dmTable in schema.Tables)
                {
                    var any = neededTables.Any(table1 => table1 == dmTable.TableName);
                    if (any) continue;

                    foreach (var relation in dmTable.ParentRelations)
                    {
                        schema.Relations.Remove(relation);
                    }

                    foreach (var relation in dmTable.ChildRelations)
                    {
                        schema.Relations.Remove(relation);
                    }

                    tablesToRemove.Add(dmTable);
                }

                foreach (var dmTable in tablesToRemove)
                {
                    schema.Tables.Remove(dmTable);
                }
            });
            _webProxyServerProvider.GetLocalProvider(HttpContext).On(interceptor);

            await _webProxyServerProvider.HandleRequestAsync(HttpContext);
        }
    }

**Comment by Mimetis** [comment]
It will be supported on the next version.
I'm currently working on it.
Most probably end of March

**Comment by mdonogma** [comment]
Thank you so much. Really appreciate the great work. This library is amazing

Sent from my iPhone

On 4 Mar 2020, at 20:48, Sébastien Pertus <notifications@github.com> wrote:

﻿

It will be supported on the next version.
I'm currently working on it.
Most probably end of March

**Comment by Mimetis** [comment]
This issue has been fixed with the release of the version `v0.5`.
If you still experiencing the same bug, please comment here, I will re-open the issue.

**Comment by mdonogma** [comment]
Love your work! Your changing the world with the stuff. It’s about time someone did a really good open source sync framework for cross platform and what you’ve achieved to date is incredible.
